### PR TITLE
Implement character play time

### PR DESCRIPTION
**Pull request by WoozChucky**
Implemented the characters played time.

TLDR: Whenever `EnterGame` packet is handled, we save the server current uptime millis in the PlayerEntity, and then when the `Logout`/`PhaseSelect` packets are handled we fetch the time again and calculate total minutes and persist that info.

### CODE REVIEW

**Comment by MeikelLP** [comment]
What about persisting it in the DB? :)

**Comment by WoozChucky** [comment]
> What about persisting it in the DB? :)

Hmm, `await _world.DespawnPlayerAsync(context.Player);` should take care of it both in `SelectPhaseHandler` and `LogoutCommand` according to the in-game test and debug that I've made. Did I missunderstood anything ?

**Comment by MeikelLP** [comment]
You have to adjust the function which maps `PlayerData` to `Player` (Entity) :)

**Comment by WoozChucky** [comment]
> You have to adjust the function which maps `PlayerData` to `Player` (Entity) :)

I'm guessing that would be in DbPlayerRepository.cs: line 76, but it is already done. Or are you refering to another location? I tried to find it 

```c#
public async Task SetPlayerAsync(PlayerData data)
    {
        var entity = await _db.Players.FirstOrDefaultAsync(x => x.Id == data.Id);
        if (entity is null) return;

        entity.Empire = data.Empire;
        entity.PlayerClass = data.PlayerClass;
        entity.SkillGroup = data.SkillGroup;
        entity.PlayTime = data.PlayTime;       // it was already mapped before my pr
        entity.Level = data.Level;
        entity.Experience = data.Experience;
        entity.Gold = data.Gold;
        entity.St = data.St;
        entity.Ht = data.Ht;
        entity.Dx = data.Dx;
        entity.Iq = data.Iq;
        entity.PositionX = data.PositionX;
        entity.PositionY = data.PositionY;
        entity.Health = data.Health;
        entity.Mana = data.Mana;
        entity.Stamina = data.Stamina;
        entity.BodyPart = data.BodyPart;
        entity.HairPart = data.HairPart;
        entity.Name = data.Name;
        entity.GivenStatusPoints = data.GivenStatusPoints;
        entity.AvailableStatusPoints = data.AvailableStatusPoints;

        await _db.SaveChangesAsync();
    }
```

**Comment by MeikelLP** [comment]
What about a forced disconnect or a disconnect because connection is lost?

**Comment by MeikelLP** [comment]
> > You have to adjust the function which maps `PlayerData` to `Player` (Entity) :)
> 
> I'm guessing that would be in DbPlayerRepository.cs: line 76, but it is already done. Or are you refering to another location? I tried to find it
> 
> ```cs
>     public async Task SetPlayerAsync(PlayerData data)
>     {
>         var entity = await _db.Players.FirstOrDefaultAsync(x => x.Id == data.Id);
>         if (entity is null) return;
> 
>         entity.Empire = data.Empire;
>         entity.PlayerClass = data.PlayerClass;
>         entity.SkillGroup = data.SkillGroup;
>         entity.PlayTime = data.PlayTime;       // it was already mapped before my pr
>         entity.Level = data.Level;
>         entity.Experience = data.Experience;
>         entity.Gold = data.Gold;
>         entity.St = data.St;
>         entity.Ht = data.Ht;
>         entity.Dx = data.Dx;
>         entity.Iq = data.Iq;
>         entity.PositionX = data.PositionX;
>         entity.PositionY = data.PositionY;
>         entity.Health = data.Health;
>         entity.Mana = data.Mana;
>         entity.Stamina = data.Stamina;
>         entity.BodyPart = data.BodyPart;
>         entity.HairPart = data.HairPart;
>         entity.Name = data.Name;
>         entity.GivenStatusPoints = data.GivenStatusPoints;
>         entity.AvailableStatusPoints = data.AvailableStatusPoints;
> 
>         await _db.SaveChangesAsync();
>     }
> ```

Oh then nevermind :)

**Comment by WoozChucky** [comment]
> What about a forced disconnect or a disconnect because connection is lost?

Implemented the code for when the connection is lost/dropped.
- Not sure if I like my own approach here because I had to change the public api of `Connection`, it's not a breaking change with existing code but still, if you have a better idea feel free to share your thoughts! :) 

In case of a forced disconnect (from either the server, or player exiting via logout/quit) the world de-spawn should take care of it.

**Comment by MeikelLP** [comment]
Sorry, another thing I thought about:
How about we track the players time in ms (long) and just display the number in minutes. This way we are much more accurate. Imagine how many minutes u lose just because of logging in and out

**Comment by WoozChucky** [comment]
> Sorry, another thing I thought about: How about we track the players time in ms (long) and just display the number in minutes. This way we are much more accurate. Imagine how many minutes u lose just because of logging in and out

Agreed!

File: src/Executables/Game/Extensions/PaketExtensions.cs [hunk 1]
```diff
@@ -13,7 +13,7 @@ public static Character ToCharacter(this PlayerData player)
             Name = player.Name,
             Class = player.PlayerClass,
             Level = player.Level,
-            Playtime = player.PlayTime,
+            Playtime = player.PlayTime == 0 ? 0 : (uint) player.PlayTime / 60000, // Milliseconds to minutes
```
Comment (MeikelLP) [inline src/Executables/Game/Extensions/PaketExtensions.cs:16]: Maybe a more readable approach would be `TimeSpan.FromMiliSeconds(player.PlayTime).TotalMinutes` but it's up to you :)

File: src/Executables/Game/World/Entities/PlayerEntity.cs [hunk 2]
```diff
@@ -676,6 +682,8 @@ public override uint GetPoint(EPoints point)
                     return _defence;
                 case EPoints.StatusPoints:
                     return Player.AvailableStatusPoints;
+                case EPoints.PlayTime:
+                    return Player.PlayTime == 0 ? 0 : (uint) Player.PlayTime / 60000; // milliseconds to minutes
```
Comment (MeikelLP) [inline src/Executables/Game/World/Entities/PlayerEntity.cs:686]: Same as above

File: src/Executables/Game/World/Entities/PlayerEntity.cs [hunk 3]
```diff
@@ -802,6 +810,17 @@ public async Task OnDespawnAsync()
             await Persist();
         }
 
+        public async Task CalculatePlayedTimeAsync()
+        {
+            var key = $"player:{Player.Id}:loggedInTime";
+            var startSessionTime = await _cacheManager.Get<long>(key);
+            var totalSessionTime = Connection.Server.ServerTime - startSessionTime;
+            //var minutes = sessionTimeMillis / 60000; // milliseconds to minutes
```
Comment (MeikelLP) [inline src/Executables/Game/World/Entities/PlayerEntity.cs:818]: Dead code?

**Review by MeikelLP — CHANGES_REQUESTED** [review]

**Comment by MeikelLP** [comment]
Pipeline is still failing. Probably due to changes in the migrations

**Comment by WoozChucky** [comment]
> Pipeline is still failing. Probably due to changes in the migrations

Yep that was exactly it, plus a test that was failing due to player fake data giving a overflow in `TimeSpan` conversion of `PlayTime`.

**Review by MeikelLP — APPROVED** [review]

**Pull request merged**